### PR TITLE
feat: add support for .mjs file output

### LINKED ISSUE #315: Extra file (package.json) not included into final zips
**Describe the bug**
Extra files (package.json to be exact) are not included into the final zips.

**Versions (please complete the following information):**
 - OS: Mac
 - Serverless Framework Version: 3.18.2
 - Plugin Version: 1.30.0

**Additional context**
So basically I wanted to use esbuild with ESM (AWS supports this for a while already). I set the `type` inside package.json to `module` and everything else needed for this move, worked fine, but the problem I stumbled upon is that serverless-esbuild simply does not include the `package.json` into the zip even if it's added to the `package.patterns` like so below.

```
custom:
  esbuild:
    bundle: true
    minify: false
    packager: 'yarn'
    target: 'node16'
    keepOutputDirectory: true

package:
  individually: true
  excludeDevDependencies: true
  patterns:
    - 'package.json'
```

It looks like the `.esbuild` contains the the `package.json`, but the zips themselves don't.

Of course I could just name my files `.mjs` instead, but that isn't an ideal solution for me.

### PR DESCRIPTION
**Pull request by codingnuclei**
Hey.

This PR adds outputting `.mjs` files if the build option `platform=neutral`.

Currently we can use `platform=neutral` to output the bundle as esm. The bundle extension however remains `.js`. This is fine when it is accompanied by a `package.json` which has `"type": "module"` set.

However, the `package.json' is explicitly not included in the final zip.

https://github.com/floydspace/serverless-esbuild/blob/master/src/pack.ts#L38

This means that lambdas run the code as CommonJs.

Changing the file type seems a more explicit option (rather than package the `package.json`)

## Considerations:
~~-`platform=neutral` now outputs `.mjs` files~~

## Notes on testing
- How typing is currently set up, it is hard to change `buildOptions` in tests.  I have therefore casted when added to the overrides.  Keeping the variable typed should ensure the type safety.

First PR to serverless-esbuild so let me know if you require anything else.

Partially closes #315 (`package.json` is no longer required)

## Update

Have updated so that `.js` is the default extension thereby making this a none breaking change.
If you wish to change the extension you can now use the new `outputFileExtension` which has some validation to keep the user right.

### CODE REVIEW

**Comment by codingnuclei** [comment]
PR changed to be more configurable

**Comment by floydspace** [comment]
Hi @codingnuclei , thank you for the careful PR. I think it's a necessary addition.
I'm happy to merge it, but please consider my comments about serverless error

**Comment by codingnuclei** [comment]
> Hi @codingnuclei , thank you for the careful PR. I think it's a necessary addition. I'm happy to merge it, but please consider my comments about serverless error

This might be me being silly - but I am not seeing any comments?

**Review by floydspace — CHANGES_REQUESTED** [review]

File: src/bundle.ts [hunk 1]
```diff
@@ -27,6 +27,20 @@ export async function bundle(this: EsbuildServerlessPlugin, incremental = false)
     plugins: this.plugins,
   };
 
+  if (
+    this.buildOptions.platform === 'neutral' &&
+    this.buildOptions.outputFileExtension === '.cjs'
+  ) {
+    throw new Error('ERROR: platform "neutral" should not output a file with extension ".cjs".');
```
[the remark this answers is not in the document]
Comment (floydspace) [reply src/bundle.ts:34]: Yes unfortunately types are not perfect, and I see the newest definitely types does not have it either,
so you could just assume it or extend types by defining custom `d.ts` for the module in the root `/typings` folder if you'd like, but it's out of the scope of this PR

[the remark this answers is not in the document]
Comment (codingnuclei) [reply src/bundle.ts]: Assumed typings for this PR

**Comment by codingnuclei** [comment]
@floydspace updated with your comments resolved :)

**Review by floydspace — APPROVED** [review]
@codingnuclei thank you, approved.

**Pull request merged**